plugins/vpc-tunnel/plugin/commands.go

```diff
@@ -0,0 +1,329 @@
+		err = geneve.SetNetNS(ns)
+		if err != nil {
+			log.Errorf("Failed to move geneve link: %v.", err)
+			return err
+		}
+	}
+
+	// Complete the remaining setup in target network namespace.
+	err = ns.Run(func() error {
+		var err error
+
+		// Create the container-facing link based on the requested interface type.
+		switch netConfig.InterfaceType {
+		case tunnelConfig.IfTypeGeneve:
+			// Container is running in a network namespace on this host.
+			err = plugin.configureGeneveLink(geneve, netConfig.IPAddresses, netConfig.GatewayIPAddress)
+		case tunnelConfig.IfTypeTAP:
+			// Container is running in a VM.
+			// Connect the geneve interface to a TAP link in the target network namespace.
+			bridgeName := fmt.Sprintf(bridgeNameFormat, netConfig.DestinationPort)
+			err = utils.CreateTAPLink(&geneve.ENI, bridgeName, args.IfName, netConfig.Tap)
+		}
+
+		// Add a blackhole route for IMDS endpoint if required.
+		if netConfig.BlockIMDS {
+			err = imds.BlockInstanceMetadataEndpoint()
+			if err != nil {
+				return err
+			}
+		}
+
+		return err
+	})
+
+	if err != nil {
+		log.Errorf("Failed to setup the link: %v.", err)
+		return err
+	}
+
+	// Generate CNI result.
+	// IP addresses, routes and DNS are configured by VPC DHCP servers.
+	result := &cniTypesCurrent.Result{
+		Interfaces: []*cniTypesCurrent.Interface{
+			{
+				Name:    args.IfName,
+				Sandbox: args.Netns,
+			},
+		},
+	}
+
+	log.Infof("Writing CNI result to stdout: %+v", result)
+
+	return cniTypes.PrintResult(result, netConfig.CNIVersion)
+}
+
+// Del is the internal implementation of CNI DEL command.
+// CNI DEL command can be called by the orchestrator agent multiple times for the same interface,
+// and thus must be best-effort and idempotent.
+func (plugin *Plugin) Del(args *cniSkel.CmdArgs) error {
+	// Parse network configuration.
+	netConfig, err := tunnelConfig.New(args)
+	if err != nil {
+		log.Errorf("Failed to parse netconfig from args: %v.", err)
+		return err
+	}
+
+	log.Infof("Executing DEL with netconfig: %+v.", netConfig)
+
+	// Derive names from CNI network config.
+	geneveName := fmt.Sprintf(GeneveLinkNameFormat, netConfig.VNI, netConfig.DestinationPort)
+	tapBridgeName := fmt.Sprintf(bridgeNameFormat, netConfig.DestinationPort)
+	tapLinkName := args.IfName
+
+	// Search for the target network namespace.
+	netns, err := netns.GetNetNS(args.Netns)
+	if err == nil {
+		// In target network namespace...
+		err = netns.Run(func() error {
+			if netConfig.InterfaceType == tunnelConfig.IfTypeTAP {
+				// Delete the tap link.
+				la := netlink.NewLinkAttrs()
+				la.Name = tapLinkName
+				tapLink := &netlink.Tuntap{LinkAttrs: la}
+				log.Infof("Deleting tap link: %v.", tapLinkName)
+				err = netlink.LinkDel(tapLink)
+				if err != nil {
+					log.Errorf("Failed to delete tap link: %v.", err)
+				}
+			}
+
+			// Delete the geneve link.
+			la := netlink.NewLinkAttrs()
+			la.Name = geneveName
+			geneveLink := &netlink.Vlan{LinkAttrs: la}
```

Comment:
May be it's a naming thing. But, the interface type can either be TAP or tunnel, right? If that's the case, is this logic applicable to TAP devices? It seems like we're deleting that twice?

Reply:
It is implemented this way so that we can bring down all network resources (GENEVE, bridge and TAP) with just one invocation of plugin. As of now, deleting in the TAP mode does not leave the network namespace and the GENEVE interface in a usable state. For that we will need to re-configure the route and ARP table entries. For current use cases, we do not have such a requirement where we need to revert back to the `geneve` setup from the `tap` setup. So it makes sense to remove all interfaces including the GENEVE interface while deleting in the TAP mode. But if there are reasons why I should not do that, I can change it.